src/backend/distributed/metadata/metadata_cache.c

```diff
@@ -4910,3 +4912,144 @@ poolinfo_valid(PG_FUNCTION_ARGS)
 							"with Citus, please contact us at "
 							"https://citusdata.com/about/contact_us")));
 }
+
+
+/*
+ * GetLocalClockValue returns the logical_clock_value value stored
+ * in the Citus catalog pg_dist_local_group.
+ */
+bool
+GetLocalClockValue(uint64 *savedValue)
+{
+	ScanKeyData scanKey[1];
+	int scanKeyCount = 0;
+	bool returnCode;
+
+	InitializeCaches();
+
+	Relation pgDistLocalGroupId = table_open(DistLocalGroupIdRelationId(),
+											 AccessShareLock);
+
+	SysScanDesc scanDescriptor = systable_beginscan(pgDistLocalGroupId,
+													InvalidOid, false,
+													NULL, scanKeyCount, scanKey);
+
+	TupleDesc tupleDescriptor = RelationGetDescr(pgDistLocalGroupId);
+
+	HeapTuple heapTuple = systable_getnext(scanDescriptor);
+
+	if (HeapTupleIsValid(heapTuple))
+	{
+		Datum datumArray[Natts_pg_dist_local_group];
+		bool isNullArray[Natts_pg_dist_local_group];
+		heap_deform_tuple(heapTuple, tupleDescriptor, datumArray, isNullArray);
+		*savedValue = DatumGetInt64(
+			datumArray[Anum_pg_dist_local_logical_clock_value - 1]);
+		returnCode = true;
+	}
+	else
+	{
+		/*
+		 * Upgrade is happening. When upgrading postgres, pg_dist_local_group is
+		 * temporarily empty before citus_finish_pg_upgrade() finishes execution.
+		 */
+		returnCode = false;
+	}
+
+	systable_endscan(scanDescriptor);
+	table_close(pgDistLocalGroupId, AccessShareLock);
+
+	return returnCode;
+}
+
+
+/*
+ * StoreLocalClockValue gets invoked in two places, one in Citus maintenance
+ * daemon, and when a ever a backend exits, but that still leaves a time
+ * window where the clock value used may not be current in the catalog.
+ * TBD: May be store it during Checkpoint too?
+ */
+void
+StoreLocalClockValue(int code, Datum arg)
+{
+	/* It's a no-op if any below are true
+	 * Dist cache(s) are not set up yet or
+	 * In the midst of creating or altering extension or
+	 * If the version checks are disabled
+	 */
+	if (IsInitProcessingMode() || creating_extension || !EnableVersionChecks ||
+		(MyBackendType != B_BACKEND && MyBackendType != B_BG_WORKER))
+	{
+		return;
+	}
+
+	StartTransactionCommand();
+
+	if (LockCitusExtension() && CheckCitusVersion(LOG) && CitusHasBeenLoaded())
+	{
+		UpdateClockCatalog();
+	}
+
+	CommitTransactionCommand();
+	CommandCounterIncrement();
+}
+
+
+/*
+ * UpdateClockCatalog persists the most recent logical clock value
+ * into the catalog pg_dist_local_group.
+ */
+static void
+UpdateClockCatalog(void)
+{
+	ScanKeyData scanKey[1];
+	int scanKeyCount = 0;
+	Datum values[Natts_pg_dist_local_group];
+	bool isNull[Natts_pg_dist_local_group];
+	bool replace[Natts_pg_dist_local_group];
```

Comment:
what about removing memset calls ?

```suggestion
	Datum values[Natts_pg_dist_local_group] = { 0 };
	bool isNull[Natts_pg_dist_local_group] = { 0 };
	bool replace[Natts_pg_dist_local_group] = { 0 };
```